composer/loggers/slack_logger.py

```diff
@@ -0,0 +1,190 @@
+        include_keys: Optional[Set[str]] = None,
+        interval_in_seconds: Optional[int] = None,
+    ) -> None:
+        self.formatter_func = formatter_func
+        self.include_keys = include_keys
+        self.interval_in_seconds = interval_in_seconds
+
+
+class SlackLogger(LoggerDestination):
+    """Logger to log metrics to slack.
+
+    This logger uses Slack's postMessage api - https://api.slack.com/methods/chat.postMessage.
+
+    Args:
+        log_metrics_config (SlackLogConfig): Configuration for logging metrics to slack.
+        log_traces_config (SlackLogConfig): Configuration for logging traces to slack.
+        log_hparams_config (SlackLogConfig): Configuration for logging hyperparameters to slack.
+    """
+
+    def __init__(
+        self,
+        log_metrics_config: Optional[SlackLogConfig] = None,
+        log_traces_config: Optional[SlackLogConfig] = None,
+        log_hparams_config: Optional[SlackLogConfig] = None,
+    ) -> None:
+        try:
+            from slack_sdk import WebClient
+            del slack_sdk
+        except ImportError as e:
+            raise MissingConditionalImportError('health_checker', 'slack_sdk', None) from e
+
+        self.client = WebClient()
+        self.slack_logging_api_key = os.environ.get('SLACK_LOGGING_API_KEY', None)
+        self.channel_id = os.environ.get('SLACK_LOGGING_CHANNEL_ID', None)
+
+        if self.slack_logging_api_key is None or self.channel_id is None:
+            raise RuntimeError(
+                'SLACK_LOGGING_API_KEY and SLACK_LOGGING_CHANNEL_ID must be set as environment variables')
+
+        self.log_metrics_config = log_metrics_config
+        self.log_traces_config = log_traces_config
+        self.log_hparams_config = log_hparams_config
+
+        # Maintain logger state
+        now = time.time()
+        self.logs_since_last_log, self.last_log_time = [], now
+        self.traces_since_last_log, self.last_trace_time = [], now
+        self.hparams_since_last_log, self.last_hparams_time = [], now
+
+    # Rich message layouts can be created using Slack Blocks Kit.
+    # See documentation here: https://api.slack.com/messaging/composing/layouts
+    def _log_to_slack(
+        self,
+        data: Dict[str, Any],
+        last_time_flushed: float,
+        blocks_to_log: List[Dict[str, Any]],
+        config: SlackLogConfig,
+        **kwargs,  # can be used to pass additional arguments to the formatter function (eg for headers)
+    ) -> Tuple[float, List[Dict[str, Any]]]:
+        filtered_data = {k: v for k, v in data.items() if k in config.include_keys} if config.include_keys else data
+        blocks = config.formatter_func(
+            filtered_data, **
+            kwargs) if config.formatter_func else self._default_log_bold_key_normal_value_pair_with_header(
+                filtered_data, **kwargs)
+
+        updated_blocks_to_log = blocks + blocks_to_log
+
+        if len(updated_blocks_to_log) > 50:
```

Comment:
should we send in groups of 50 rather than just erroring out? or at least truncate and warn rather than error. I would not be super happy if the slack logger taking in too many messages crashed my run :P

Reply:
Ill make the change to flush the logs when exceeding 50. 

Was probably also a mistake to pick `RuntimeError` for these types of errors. Havent looked into the error boundaries in trainer, but wondering if we've thought of having a `LoggerError` class that could be caught and printed to std out rather than throwing?

Reply:
we could probably be better with our errors. I think we mostly just abuse the standard errors and don't do much special try/catch like that